### PR TITLE
[546] Coverage tests for Chain Traces + refactor Deleg Trace coverage…

### LINKED ISSUE #546: Add coverage tests for CHAIN traces
We need to make sure we're covering important test cases in our conformance tests. For instance:

- How many epoch changes are we generating?
- How many protocol version are we generating?
- What is the delegation churn? (i.e. how many times a given key changes
      its delegate)

For the coverage tests we can also take the coverage tests defined for the
ledger, and **adapt them** to the `CHAIN` transition system. Adaptation is
needed since we do not want to generate the same number of delegation
certificates or update proposals as in the ledger tests.

See also https://github.com/input-output-hk/cardano-ledger-specs/issues/297

### PR DESCRIPTION
**Pull request by uroboros**
Closes #546 

chain trace coverage

* based the trace coverage tests for Delegation by @dnadales, this PR adapts these coverage tests to Chain Traces
* also refactors the delegation trace tests to include some insights from this PR
* adds coverage tests for: change of delegation, repeated delegations, distribution of certificates over blocks, epoch boundaries

NOTE some coverage properties happen quite rarely given current generators - is it worth addressing this?

* "some blocks have 5 or more certificates" (only true for 10% of traces)
* "not too many delegations (at most 25%) are repeats" (only 20%)
* "not too many delegators (at most 25%) change their delegation" (only 8%)
* "not too many certificates (at most 25%) self-delegate" (only 10%)
* "the majority of blocks (at least 75%) have certificates" (only 5%)
* "there are more certificates than blocks" (only 15%)

### CODE REVIEW

**Comment by dnadales** [comment]
I appreciate the screenshot in the description @uroboros :)

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
 import           Data.Ord (Down (Down), comparing)
-import           Hedgehog (MonadTest, Property, assert, failure, forAll, property, withTests, (===))
+
+import           Hedgehog (MonadTest, Property, assert, cover, failure, forAll, property, withTests,
+                           (===))
 
 import           Control.State.Transition
 import           Control.State.Transition.Generator (TraceLength (Maximum), classifyTraceLength,
-                     traceSigGen)
+                                                     traceSigGen)
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: Check that you have the `master` version of `stylysh-haskell` installed, and that it is using our configuration.

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs [hunk 2]
```diff
@@ -72,3 +79,110 @@ signersListIsBoundedByK = property $ do
...
+        "there are more certificates than blocks"
+        (traceLength tr <= length certs)
+
+  -- for at least 5% of traces...
+  cover 5
+        "the majority of blocks (at least 75%) have certificates"
+        (emptyDelegationPayloadRatio (traceDCertsByBlock tr) <= 0.25)
+
+  -- for at least 50% of traces...
+  cover 50
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: This is really weird, I would expect a higher percentage of delegations to this epoch. I'm thinking we might want to make an issue to investigate this ...

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs [hunk 3]
```diff
@@ -72,3 +79,110 @@ signersListIsBoundedByK = property $ do
...
+  let certs = traceDCerts tr
+
+  -- for at least 15% of traces...
+  cover 15
+        "there are more certificates than blocks"
+        (traceLength tr <= length certs)
+
+  -- for at least 5% of traces...
+  cover 5
+        "the majority of blocks (at least 75%) have certificates"
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: Would it be possible to simplify this and the following sentences?  What about:

```suggestion
        "at least 75% of blocks have certificates"
```

Comment (dnadales) [reply byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: And would it make sense :)

Comment (dnadales) [reply byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: I prefer simple (and precise) English whenever possible ;)

**Comment by dnadales** [comment]
> NOTE some coverage properties happen quite rarely given current generators - is it worth addressing this?
> 
It's hard to say. Ideally we would like to identify first the fact that we're not seeing enough coverage of a particular case, and then work to improve the coverage metrics. At the moment might be too early to tell if `x%` is enough or not :man_shrugging: 

> * "some blocks have 5 or more certificates" (only true for 10% of traces)
A potential source of confusion might be the fact that we use `Maximum` instead of `Desired` for specifying the trace length. This makes it a bit hard to reason about coverage figures. If I change the `Maximum` to `Desired` when the coverage is about 30%. But all in all I would say that 5 certificates per block is a lot. 

> * "not too many delegations (at most 25%) are repeats" (only 20%)
I'd say this looks ok. Repeating delegations (delegating to the same key) is a very interesting test case.

> * "not too many delegators (at most 25%) change their delegation" (only 8%)
Again, it is good that many delegations change their delegations :) 

> * "not too many certificates (at most 25%) self-delegate" (only 10%)
Hmmmm, I'm starting to think I'm missing something :) I'd say that self-delegation is an interesting case, and we want to test this.

> * "the majority of blocks (at least 75%) have certificates" (only 5%)
Oh, this is also fine: we can't expect to have 75% of blocks with delegation certificates since as soon as all the genesis keys delegate to this or the next epoch no new delegation certs can be issued.

> * "there are more certificates than blocks" (only 15%)
This is fine I guess. Specially when we start generating delegation certificates sporadically.

So all in all, this is a good first step, and we might need to revisit the coverage figures later on.

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs [hunk 4]
```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
...
+--------------------------------------------------------------------------------
+
+delegatorDelegate :: DCert -> (VKeyGenesis, VKey)
+delegatorDelegate = delegator &&& delegate
+
+ratioInt :: Int -> Int -> Double
+ratioInt x y
+  = fromIntegral x / fromIntegral y
+
+-- | Filters the list with the predicate and returns
```
Comment (dnadales) [inline byron/ledger/executable-spec/src/Ledger/Delegation.hs]: Hmm, this comment doesn't seem right ...

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs [hunk 5]
```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
...
+
+-- | True if the tuple count (snd item) is >= 2
+multiple :: (a,Int) -> Bool
+multiple = (2 <=) . snd
+
+-- | Count the number of delegates in the given certificates
+delegateCounts :: [DCert] -> [(VKey, Int)]
+delegateCounts certs
+  = fmap delegatorDelegate certs
+    -- Remove duplicated elements, since we're not
```
Comment (dnadales) [inline byron/ledger/executable-spec/src/Ledger/Delegation.hs]: BTW, I think we can expand the line length of the comments (if you're using Emacs you can reformat this with `M-q`)

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs [hunk 6]
```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
...
+    & List.nub
+    -- Select the (unique) delegates
+    & fmap snd
+    -- If we have more than one occurrence of a delegate,
+    -- then we know that there were multiple delegations
+    -- to that delegate
+    & count
+
+-- | Count the number of delegators in the given certificates
+delegatorCounts :: [DCert] -> [(VKeyGenesis, Int)]
```
Comment (mdimjasevic) [inline byron/ledger/executable-spec/src/Ledger/Delegation.hs]: The implementation of `delegatorCounts` is almost identical to that of `delegateCounts`. How about making an abstract function that based on an input argument either `fmap fst`s or `fmap snd`s and then make the two count functions call into that abstract function?

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs [hunk 7]
```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
...
+-- | Ratio of delegates that have multiple delegators
+-- that are delegating to them
+multipleDelegationsRatio :: [DCert] -> Double
+multipleDelegationsRatio dcerts
+  = lenRatio (filter multiple) (delegateCounts dcerts)
+
+-- | The maximum number of delegators to any particular delegate
+maxDelegationsTo :: [DCert] -> Int
+maxDelegationsTo dcerts
+  =  maxInt (snd <$> filter multiple (delegateCounts dcerts))
```
Comment (mdimjasevic) [inline byron/ledger/executable-spec/src/Ledger/Delegation.hs]: What does this give in case everyone has exactly one delegator?

File: byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs [hunk 8]
```diff
@@ -54,6 +54,6 @@ relevantKValuesAreGenerated = withTests 500 $ property $ do
     --
     -- And if we round this value down we get 100 epochs.
 
-    cover 10
+    cover 6
```
Comment (mdimjasevic) [inline byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs]: I'd argue it'd be useful to leave a comment next to changes like this one.

File: byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs [hunk 9]
```diff
@@ -325,88 +331,50 @@ dblockTracesAreClassified = withTests 200 $ property $ do
...
 relevantCasesAreCovered :: Property
 relevantCasesAreCovered = withTests 400 $ property $ do
   let tl = 1000
   tr <- forAll (trace @DBLOCK tl)
 
-  -- 70% of the traces must contain are as many delegation certificates as
-  -- blocks.
-  cover 50
+  -- 40% of the traces must contain as many delegation certificates as blocks.
+  cover 40
```
Comment (mdimjasevic) [inline byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs]: Can you explain why this change? Were generators modified that begged for updating these thresholds?

**Comment by dnadales** [comment]
> The thresholds in coverage properties seem arbitrary, but this is an observation not specific to this pull request; instead it applies to coverage properties in general.

I agree. I find it quite difficult to come with non-arbitrary thresholds. If you have any suggestions they'll be very welcome by @uroboros and myself :pray:

**Comment by dnadales** [comment]
> How about moving all coverage criteria to properties they are relevant to? @dnadales, I find that kind of organization of coverage criteria more useful than bundling all coverage criteria into an artificial and isolated Property.

Hmm, I don't know. I guess it depends on the specific case. If we factor out all the coverage checks into separate properties we'll have a lot of duplication. But feel free to adapt this in future PR's if you think that's sensible.

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs (hunk 7 above)
Comment (dnadales) [reply byron/ledger/executable-spec/src/Ledger/Delegation.hs]: Good catch. It seems this will return 0 in that case, which is not right :)

File: byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs (hunk 9 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs]: The threshold lied very close to the values that we got in the runs. Like you noticed, these thresholds are arbitrary, so these kind of adaptations would be needed. But indeed, we should ask for explanations whenever we change these :)

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs (hunk 2 above)
Comment (uroboros) [reply byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: will do

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs (hunk 6 above)
Comment (uroboros) [reply byron/ledger/executable-spec/src/Ledger/Delegation.hs]: good point... done

File: byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs (hunk 8 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs]: (this change was unrelated to this PR, it was a setting that was cutting it too fine and happened to blow up now)

File: byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs (hunk 9 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs]: these settings are still a bit brittle since we're still changing Chain trace generation, it will stabilise in due

File: byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs (hunk 8 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs]: My point is that we're lacking understanding why "cover 10" doesn't work and why lovering it to 6 works. It appears to be a black box, yet it shouldn't: try out a number and if it works, go with it. Again, this is not necessarily limited to your PR, but it could apply to coverage testing in general.

File: byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs [hunk 10]
```diff
@@ -325,88 +331,50 @@ dblockTracesAreClassified = withTests 200 $ property $ do
...
   -- epoch.
-  cover 50
+  cover 70
         "at least 50% of the certificates delegate in the next epoch"
-        (0.5 <= ratio nextEpochDelegations tr)
+        (0.5 <= nextEpochDelegationsRatio (epochDelegationEpoch tr))
 
-  -- 80% of the traces must contain at least 30% of self-delegations.
-  cover 80
+  -- 30% of the traces must contain at least 30% of self-delegations.
```
[the remark this answers is not in the document]
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs]: this is an interesting effect of the re-phrasing of these ratio's we discussed: the ratio is now the number of self-delegs to the total number of certs, whereas before it was the ratio of self-delegs to blocks in the chain

File: byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs (hunk 8 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs]: it's true, these numbers are arbitrary and not exact, but I think they do give an indication of the shape and quality of traces and they tell us that our generation is matching with our testing to some extent at least

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs (hunk 1 above)
Comment (uroboros) [reply byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: oops, done

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs [hunk 11]
```diff
@@ -78,3 +85,110 @@ signersListIsBoundedByK = property $ do
...
+  cover 60
+        "at least 25% delegates have multiple delegators"
+        (0.25 <= multipleDelegationsRatio certs)
+
+  -- for at least 20% of traces...
+  cover 20
+        "some delegates have at least 5 corresponding delegators"
+        (5 <= maxDelegationsTo certs)
+
+  -- for at least ?% of traces...
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: Question mark remaining.

File: byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs (hunk 8 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Core/Generators/Properties.hs]: >  try out a number and if it works, go with it. Again, this is not necessarily limited to your PR, but it could apply to coverage testing in general.

The way I did coverage test was to:

- write a coverage test
- check the figures we got out
  - if they made sense then lower the threshold by 10% to make sure we don't have brittle tests.
  - if they didn't make sense (for instance only 1% of self delegations), check the generators, and run the tests again.

Is this optimal? No. Are we aware of this? Yes. This is what we have at the moment, and what we could achieve in the time and resources we had. Rest assured that some of us are putting a lot of effort into this, and sometimes putting our personal time as well. Improvement proposals are always welcome.

File: byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs]: Hmmm, but in this case I would expect the ratio to go up (since we would have a smaller denominator) :thinking: 

But anyway, we can leave this for now and give it another look later on.

File: byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs [hunk 12]
```diff
@@ -78,3 +85,110 @@ signersListIsBoundedByK = property $ do
...
+        "at least 25% of delegates will delegate to this epoch"
+        (0.25 <= thisEpochDelegationsRatio (epochDelegationEpoch tr))
+
+  -- for at least 60% of traces...
+  cover 60
+        "at least 50% of delegations will delegate to the next epoch"
+        (0.5 <= nextEpochDelegationsRatio (epochDelegationEpoch tr))
+
+  -- for at least 20% of traces...
+  cover 20
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/Spec/Chain/STS/Properties.hs]: In the test I ran this results in 20% of coverage, we might want to either change this to 10% or change the threshold below to `.25`.
